File: bin/src/models/search_result.dart
import 'package:json_annotation/json_annotation.dart';

part 'search_result.g.dart';

@JsonSerializable()
class SearchResult {
  const SearchResult({
    required this.objectID,
    required this.emoji,
    required this.entity,
    required this.code,
    required this.description,
    required this.name,
    this.semver,
  });

  final String objectID;
  final String emoji;
  final String entity;
  final String code;
  final String description;
  final String name;
  final String? semver;

  static const List<String> attributesToRetrieve = [
    'emoji',
    'entity',
    'code',
    'description',
    'name',
    'semver',
  ];

  factory SearchResult.fromJson(Map<String, dynamic> json) =>
      _$SearchResultFromJson(json);

  Map<String, dynamic> toJson() => _$SearchResultToJson(this);
}

File: src/models/search_result.rs
use serde::Deserialize;

/// A Gitmoji result returned by the Algolia search index.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
pub struct SearchResult {
    /// Algolia object identifier.
    #[serde(rename = "objectID")]
    pub object_id: String,
    /// Unicode emoji value.
    pub emoji: String,
    /// HTML entity representation.
    pub entity: String,
    /// Gitmoji code such as :bug:.
    pub code: String,
    /// Human-readable result description.
    pub description: String,
    /// Gitmoji name.
    pub name: String,
    /// Optional Gitmoji semantic version.
    pub semver: Option<String>,
}

/// Minimal subset of the Algolia single-index search response.
#[derive(Debug, Deserialize)]
pub struct SearchResponse {
    /// Search results in provider-defined ranking order.
    pub hits: Vec<SearchResult>,
}
